nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/factory/V3SNMPFactory.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.snmp.factory;
+
+import org.apache.nifi.snmp.configuration.SNMPConfiguration;
+import org.apache.nifi.snmp.utils.SNMPUtils;
+import org.snmp4j.Snmp;
+import org.snmp4j.Target;
+import org.snmp4j.mp.MPv3;
+import org.snmp4j.mp.SnmpConstants;
+import org.snmp4j.security.SecurityModels;
+import org.snmp4j.security.SecurityProtocols;
+import org.snmp4j.security.USM;
+import org.snmp4j.security.UsmUser;
+import org.snmp4j.smi.OctetString;
+
+public class V3SNMPFactory extends AbstractSNMPFactory implements SNMPFactory {
+
+    @Override
+    public boolean supports(int version) {
+        return SnmpConstants.version3 == version;
+    }
+
+    @Override
+    public Snmp createSnmpManagerInstance(SNMPConfiguration configuration) {
+        final Snmp snmp = createSimpleSnmpClient();
+
+        // If there's a USM instance associated with the MPv3 bound to this Snmp instance (like an agent running
+        // on the same host) it is not null.
+        if (snmp.getUSM() == null) {
+            final OctetString localEngineId = new OctetString(MPv3.createLocalEngineID());
+            final USM usm = new USM(SecurityProtocols.getInstance(), localEngineId, 0);
+            SecurityModels.getInstance().addSecurityModel(usm);
+        }
+
+        final String username = configuration.getSecurityName();
+        final String authProtocol = configuration.getAuthProtocol();
+        final String authPassword = configuration.getAuthPassword();
+        final String privacyProtocol = configuration.getPrivacyProtocol();
+        final String privacyPassword = configuration.getPrivacyPassword();
+        final OctetString authPasswordOctet = authPassword != null ? new OctetString(authPassword) : null;
+        final OctetString privacyPasswordOctet = privacyPassword != null ? new OctetString(privacyPassword) : null;
+
+        // Add user information.
+        snmp.getUSM().addUser(
```

Comment:
It seems SNMP4J library stores the authenticated users in static context in the JVM.
Eg. when I configure a processor instance with valid user credentials and start it up, then other processors using the same username but invalid credentials can also authenticate to the agent and run without error.
To avoid this interference, `@RequiresInstanceClassLoading` may need to be used.